### PR TITLE
feat: implement most API endpoints, sans photo upload

### PR DESCRIPTION
**Pull request by kanru**

### CODE REVIEW

**Comment by kanru** [comment]
Command for testing file uploads:

```sh
curl localhost:8080/graphql \
  -F operations='{ "query": "mutation ($file: Upload!) { uploadDishPhoto(file: $file) { success filename } }", "variables": { "file": null } }' \
  -F map='{ "0": ["variables.file"] }' \
  -F 0=@photo.jpg
```

File: index.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+const typeDefs = gql`
+    type Dish {
+        id: ID!
+        name: String!
+        meal: Meal!
+        ingredient: [RecipeIngredient]!
+        meat: Float
+        vegetable: Float
+        seafood: Float
```
Comment (evelynhung) [inline index.js]: Let's remove these. They are not in use and I guess we can infer them from the ingredient.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        spice: Boolean
+        cook_time: Int
+        iCook: String
```
Comment (evelynhung) [inline index.js]: `link` or `recipe_link` is more general.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        amount: Int
+        solo(is: Boolean): Boolean
```
Comment (evelynhung) [inline index.js]: Let's use `is_one_dish` or `one_dish`

File: index.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        name: String!
+        meal: Meal!
+        ingredient: [RecipeIngredient]!
+        meat: Float
+        vegetable: Float
+        seafood: Float
+        spice: Boolean
+        cook_time: Int
+        iCook: String
+        amount: Int
```
Comment (evelynhung) [inline index.js]: Let's use `serving`.

File: index.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+async function startApolloServer(typeDefs, resolvers) {
+    const app = express();
+    const httpServer = http.createServer(app);
+    const dbusername = encodeURIComponent(process.env.DB_USERNAME);
+    const dbpassword = encodeURIComponent(process.env.DB_PASSWORD);
+    const clusterUrl = process.env.DB_URL;
+    const client = new MongoClient(`mongodb://${dbusername}:${dbpassword}@${clusterUrl}/`);
+    await client.connect();
+    const db = client.db("test");
```
Comment (evelynhung) [inline index.js]: let's read db name from a env variable.

File: index.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    type Query {
+        dishes: [Dish]
+        ingredients: [Ingredient]
+    }
+
+    type Mutation {
+        shareWeeklyMenu(menu: ShareableWeeklyMenuInput): ShareableWeeklyMenu
+        uploadDishPhoto(file: Upload!): File!
```
Comment (evelynhung) [inline index.js]: We don't need to upload photo alone. Photo is always updated with dishes.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        encoding: String
+    }
+
+    type Query {
+        dishes: [Dish]
+        ingredients: [Ingredient]
+    }
+
+    type Mutation {
+        shareWeeklyMenu(menu: ShareableWeeklyMenuInput): ShareableWeeklyMenu
```
Comment (evelynhung) [inline index.js]: The menu may not be "weekly". I am thinking to have users specifying date ranges...

File: index.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        meal: Meal!
+        ingredient: [NewDishRecipeIngredientInput]!
+        meat: Float
+        vegetable: Float
+        seafood: Float
+        spice: Boolean
+        cook_time: Int
+        iCook: String
+        amount: Int
+        solo: Boolean
```
Comment (evelynhung) [inline index.js]: The above `Dish` changes apply to here too.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+const typeDefs = gql`
+    type Dish {
+        id: ID!
+        name: String!
+        meal: Meal!
+        ingredient: [RecipeIngredient]!
+        meat: Float
+        vegetable: Float
+        seafood: Float
+        spice: Boolean
```
Comment (evelynhung) [inline index.js]: `spicy` seems better. I am thinking this could be a `Float` type.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    input NewDishRecipeIngredientInput {
+        description: String!
+        quantity: String!
+    }
+
+    input UpdateDishRecipeIngredientInput {
+        description: String!
+        quantity: String!
+    }
```
Comment (evelynhung) [inline index.js]: Could we merge `NewDishRecipeIngredientInput ` and `UpdateDishRecipeIngredientInput` and let it be `DishRecipeIngredientInput `?

Also replace `description: String!` with `ingredient: String!`  where the ingredient String is mapping to the name of the ingredient.

File: index.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+import { GraphQLUpload, graphqlUploadExpress } from 'graphql-upload';
+import { MongoClient } from 'mongodb';
+import express from 'express';
+import http from 'http';
+import Dishes from './src/data-sources/Dishes.js';
+import Ingredients from './src/data-sources/Ingredients.js';
+import Shares from './src/data-sources/Shares.js';
+import Photos from './src/data-sources/Photos.js';
+
+const typeDefs = gql`
```
Comment (evelynhung) [inline index.js]: Could we move `typeDefs` and `resolvers` to a separate file?

File: index.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        ingredients: [Ingredient]
+    }
+
+    type Mutation {
+        shareWeeklyMenu(menu: ShareableWeeklyMenuInput): ShareableWeeklyMenu
+        uploadDishPhoto(file: Upload!): File!
+        addNewDishes(dishes: [NewDishInput]): MutationResult
+        updateDishes(dishes: [UpdateDishInput]): MutationResult
+        addNewIngredients(ingredients: [NewIngredientInput]): MutationResult
+        updateIngredients(ingredients: [UpdateIngredientInput]): MutationResult
```
Comment (evelynhung) [inline index.js]: I haven't thought of a usage of this API, maybe we add it later?

File: index.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        style: String
+    }
+
+    enum Meal {
+        lunch
+        dinner
+    }
+
+    type RecipeIngredient {
+        description: String!
```
Comment (evelynhung) [inline index.js]: Let's use `ingredient: Ingredient!` so it can be linked to the ingredient instance.

File: src/data-sources/Photos.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        const filename = `${key}.${extension}`;
+        const path = `${this.base_path}/${filename}`
+        const out = createWriteStream(path);
+        stream.pipe(out);
+        stream.on('error', (e) => {
+            out.destroy(e);
+            return { success: false };
+        })
+        await new Promise(resolve => out.on('finish', resolve));
+        await this.collection.insertOne({ key, filename, mimetype, encoding });
```
Comment (evelynhung) [inline src/data-sources/Photos.js]: I would like to associate this photo to a dish as 1-1 mapping.

**Review by evelynhung — COMMENTED** [review]
:D nice progress!

**Review by evelynhung — COMMENTED** [review]
Serve the dish! :D

**Review by evelynhung — APPROVED** [review]

**Pull request merged**